### PR TITLE
sketch-map

### PR DESCRIPTION
**Pull request by sofiia-chorna**
re-implementation of sketch-map (https://sketchmap.org/) from c++ code (https://github.com/cosmo-epfl/sketchmap.git).

added `decomposition.SketchMap`, `sample_selection.voronoi_weights`, `datasets.load_sketchmap_dimred_reference` to the public API and a doc with an example

for validation, i reproduced the protein example from the c++ repo and did a separate projection from scratch to compare results (all in the example doc)

for the global-optimization stage there are two options:
`global_optimizer="gradient"` => L-BFGS, fast and works for any `n_components`
`global_optimizer="grid"` => same as in the c++ - in 2d pointwise sweeps with anneling, slower but reaches the same result as c++ 

as the stress is non-convex, the two optimizers can land in slightly different minima

in a separate PR will add an optional torch backend which i used to fit sketch-map on >1m points for our MAD-CORE draft

i cannot request reviewers on this repo, so pinging instead ^^
@PicoCentauri and @rosecers for the scikit-matter side and probably @ceriottm as the author of the original sketch-map 😁

Contributor (creator of PR) checklist
-------------------------------------
 - [x] Tests updated (for new features and bugfixes)?
 - [x] Documentation updated (for new features)?
 - [ ] Issue referenced (for PRs that solve an issue)?

For Reviewer
------------
 - [ ] CHANGELOG updated if important change?

----
📚 Documentation preview 📚: https://scikit-matter--284.org.readthedocs.build/en/284/

### CODE REVIEW

File: docs/src/figures/protein_maps.png
Comment (PicoCentauri) [inline docs/src/figures/protein_maps.png:1]: can we do this as svg?

File: docs/src/references/decomposition.rst [hunk 1]
```diff
@@ -70,3 +70,27 @@ Kernel PCovC
...
+simulations, where configurations cluster into basins. Short distances mostly measure
+thermal fluctuations inside a basin, long ones say little more than that two
+configurations are far apart, so the structure worth preserving lives in between.
+sketch-map therefore passes the high- and the low-dimensional distances through a
+saturating sigmoid before matching them, and the embedding is driven by the intermediate
+range.
+
+The :ref:`example <sphx_glr_examples_decomposition_sketchmap.py>` introduces the method,
+walks through the landmark workflow used for large datasets, and validates the
+implementation against the reference C++ code.
```
Comment (PicoCentauri) [inline docs/src/references/decomposition.rst]: I would say validation we don't need in an example. Can be in the tests though.

File: examples/decomposition/sketchmap.py [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+
+r"""
+sketch-map: nonlinear dimensionality reduction
+==============================================
+
+sketch-map [Ceriotti2011]_ is a nonlinear dimensionality-reduction algorithm for
+atomistic simulations, where configurations cluster into basins. Only intermediate
+distances are informative there, as short ones measure thermal fluctuations inside a
+basin and long ones only say that two configurations lie in different basins. Methods
+that reproduce all distances equally, like MDS, or maximize retained variance, like PCA,
```
Comment (PicoCentauri) [inline examples/decomposition/sketchmap.py]: maybe link MDS and PCA to corresponding sklearn classes/examples.

File: examples/decomposition/sketchmap.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+landmarks = X[fps.selected_idx_]
+landmark_basin = basin[fps.selected_idx_]
+weights = voronoi_weights(X, landmarks)
+
+# %%
+# Letting sketch-map choose its parameters
+# ----------------------------------------
+#
+# All five sigmoid parameters default to ``None``, in which case they are estimated from
+# the data, so ``SketchMap()`` can be fitted without tuning anything. The values it
```
Comment (PicoCentauri) [inline examples/decomposition/sketchmap.py]:
```suggestion
# the data, so :class:`SketchMap` can be fitted without tuning anything. The values it
```

File: examples/decomposition/sketchmap.py [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+)
+ax.set_xlabel("SMAP1")
+ax.set_ylabel("SMAP2")
+ax.set_aspect("equal", "box")
+fig.tight_layout()
+
+
+# %%
+# Validating against the reference C++ implementation
+# ---------------------------------------------------
```
Comment (PicoCentauri) [inline examples/decomposition/sketchmap.py]: I think this can be removed from the example.

File: examples/decomposition/sketchmap.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
+#!/usr/bin/env python
+# coding: utf-8
+
```
Comment (PicoCentauri) [inline examples/decomposition/sketchmap.py:3]: can we use the second figure as thumbnail? It looks nicer



```suggestion

# sphinx_gallery_thumbnail_number = 2
```

File: tests/data/sketchmap_dimred_reference.dat [hunk 6]
```diff
@@ -0,0 +1,65 @@
+# Columns: y1 y2. Reported ld_error (stress): 0.0129757
```
Comment (PicoCentauri) [inline tests/data/sketchmap_dimred_reference.dat:4]: Maybe also add here that this comes from the C++ code. I know we have the readme but a comment inside the file is always good.

File: src/skmatter/decomposition/_sketchmap_utils.py [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+
+    return coordinates
+
+
+def _gaussian(x, amplitude, center, std_dev):
+    """Gaussian function for curve fitting in distance distribution analysis"""
+    return amplitude * np.exp(-((x - center) ** 2) / (2 * std_dev**2))
+
+
+def _maybe_tqdm(iterable, enabled, **tqdm_kwargs):
```
Comment (PicoCentauri) [inline src/skmatter/decomposition/_sketchmap_utils.py]: Isn't there already a function for this? Or do we always hardcode this?

**Review by PicoCentauri — COMMENTED** [review]
Very nice that this is working @sofiia-chorna.

The only bigger change is that I would remove the C++ dataset and also don't compare in the example. Should be enough that we do a regression test.

File: docs/src/figures/protein_maps.png
Comment (sofiia-chorna) [reply docs/src/figures/protein_maps.png:1]: done in https://github.com/scikit-learn-contrib/scikit-matter/pull/284/commits/4c31bc6ca89a1a81fa08b0713cefbbcb9c0b6d0d

File: examples/decomposition/sketchmap.py (hunk 5 above)
Comment (sofiia-chorna) [reply examples/decomposition/sketchmap.py:3]: done https://github.com/scikit-learn-contrib/scikit-matter/pull/284/commits/250ad6aac5e7ab0e2b2e4df9f9b01e8998199a3d

File: examples/decomposition/sketchmap.py (hunk 2 above)
Comment (sofiia-chorna) [reply examples/decomposition/sketchmap.py]: done https://github.com/scikit-learn-contrib/scikit-matter/pull/284/commits/250ad6aac5e7ab0e2b2e4df9f9b01e8998199a3d

File: docs/src/references/decomposition.rst (hunk 1 above)
Comment (sofiia-chorna) [reply docs/src/references/decomposition.rst]: okay, moved the executable part to the tests
in https://github.com/scikit-learn-contrib/scikit-matter/pull/284/changes/dcb3aa289bcbcf558974ed23d74a632790324111 and https://github.com/scikit-learn-contrib/scikit-matter/pull/284/changes/3bc438981e2ec9bef29ddd63cdd52b65b227a361

File: tests/data/sketchmap_dimred_reference.dat (hunk 6 above)
Comment (sofiia-chorna) [reply tests/data/sketchmap_dimred_reference.dat:4]: done! https://github.com/scikit-learn-contrib/scikit-matter/commit/21fd873cb9807c59a4fe33f82bf4184f092fcb9c

File: src/skmatter/decomposition/_sketchmap_utils.py (hunk 7 above)
Comment (sofiia-chorna) [reply src/skmatter/decomposition/_sketchmap_utils.py]: fixed! https://github.com/scikit-learn-contrib/scikit-matter/commit/21fd873cb9807c59a4fe33f82bf4184f092fcb9c

**Comment by sofiia-chorna** [comment]
@PicoCentauri thanks a lot for the review 🙏

i removed the digits comparison from the example & moved the frozen c++ embedding to tests/data/ and to the test

the example keeps only a part with the full reproduction of the published smap example since executing it in the CI would take 13 minutes!

**Review by PicoCentauri — APPROVED** [review]
Nice, lets merge this and do a release. Thanks @sofiia-chorna
